jupyterhub/files/hub/jupyterhub_config.py

```diff
@@ -107,28 +107,38 @@ def camelCaseify(s):
 )
 
 # implement common labels
-# this duplicates the jupyterhub.commonLabels helper
+# This mimics the jupyterhub.commonLabels helper, but declares managed-by to
+# kubespawner instead of helm.
+#
+# The labels app and release are old labels enabled to be deleted in z2jh 5, but
+# for now retained to avoid a breaking change in z2jh 4 that would force user
+# server restarts. Restarts would be required because NetworkPolicy resources
+# must select old/new pods with labels that then needs to be seen on both
+# old/new pods, and we want these resources to keep functioning for old/new user
+# server pods during an upgrade.
+#
 common_labels = c.KubeSpawner.common_labels = {}
-common_labels["app"] = get_config(
+common_labels["app.kubernetes.io/name"] = common_labels["app"] = get_config(
     "nameOverride",
     default=get_config("Chart.Name", "jupyterhub"),
 )
-common_labels["heritage"] = "jupyterhub"
+release = get_config("Release.Name")
+if release:
+    common_labels["app.kubernetes.io/instance"] = common_labels["release"] = release
 chart_name = get_config("Chart.Name")
 chart_version = get_config("Chart.Version")
 if chart_name and chart_version:
-    common_labels["chart"] = "{}-{}".format(
-        chart_name,
-        chart_version.replace("+", "_"),
+    common_labels["helm.sh/chart"] = common_labels["chart"] = (
+        f"{chart_name}-{chart_version.replace('+', '_')}"
     )
-release = get_config("Release.Name")
-if release:
-    common_labels["release"] = release
+chart_app_version = get_config("Chart.AppVersion")
+if chart_app_version:
+    common_labels["app.kubernetes.io/version"] = chart_app_version
```

Comment:
I'm not sure about this. The app version in the singleuser pods will often be different to the app version in the hub. Is it still useful to have the hub version on the singleuser pods?

Reply:
To know a version when kubespawner created resource could be relevant, but we still could get this from helm.sh/chart that includes the chart version. For example if we document an instruction that requires restarts of user servers started with z2jh/jupyterhub/kubespawner of some version that was too old, then a z2jh/jupyterhub/kubespawner version field could help us not restart all user servers but only some.

But, app.kubernetes.io/version for a kubespawner created resource - should that be version of jupyterhub server, jupyterhub inside user server, or kubespawner? The ambiguity for this situation makes it less valuable to provide a version at all.

- I think if we have version set, it should be the jupyterhub version deployed by the helm chart in hub pod, I believe that [matches the labels intent best according to k8s docs](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/#labels) - at least if we have `app.kubernetes.io/name=jupyterhub`.
- If we'd set `app.kubernetes.io/name=kubespawner` then the version should be set to the version of kubespawner, and we should use `app.kubernetes.io/part-of=jupyterhub`

---

Messy...

I think I lean towards either retaining version as it is, or removing it atm.

I've consiered also changing the .../name label to kubespawner, but I'm hesitant towards doing it as searching for labels .../name=jupyterhub .../instance=my-helm-release can successfully identify all chart resources - directly created or indirectly via kubespawner, while also allowing for filtering to those managed by helm and kubespawner via the .../managed-by label.

@manics lets make a call - do we remove it? I think whatever we do, its not a big decision and its fine to do either.

Reply:
> The metadata is organized around the concept of an application. Kubernetes is not a platform as a service (PaaS) and doesn't have or enforce a formal notion of an application. Instead, applications are informal and described with metadata. The definition of what an application contains is loose.

The problem is we don't have a single "application". Intuitively I think treating `version` as the `component` version is helpful, but since that's going to be too complicated maybe we remove it for now, or just add it to the hub pod only?

Reply:
I systematically removed it from all chart resources and kubespawner resources created in 5099dfc!